Add predefined datatypes for bfloat16 data

jhendersonHDF opened the pull request:
Adds predefined datatypes for little- and big-endian bfloat16 data

Does not add support for any native bfloat16 types; datatype conversions are performed in software

Also adds missing float16 predefined types to fortran
<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Add predefined datatypes for bfloat16 in little- and big-endian formats, and update Fortran with missing float16 types.
> 
>   - **Behavior**:
>     - Adds predefined datatypes `FLOAT_BFLOAT16BE` and `FLOAT_BFLOAT16LE` in `H5PredType.cpp` and `H5PredType.h`.
>     - Adds missing `float16` types to Fortran in `H5_f.c`, `H5_ff.F90`, and `H5f90global.F90`.
>     - Updates `h5init_types_c()` in `H5_f.c` to include new bfloat16 and float16 types.
>   - **Tests**:
>     - Adds tests for bfloat16 special values in `test_bfloat16()` in `dt_arith.c`.
>     - Adds tests for bfloat16 datatype characteristics in `test_bfloat16()` in `dtypes.c`.
>     - Updates `h5dumpgentest.c` to generate test files for bfloat16.
>   - **Misc**:
>     - Updates `HDF5Constants.java` and `h5util.c` for Java support of bfloat16.
>     - Updates `H5T.c` and `H5Tnative.c` to initialize and handle bfloat16 types.
>     - Updates `h5tools_type.c` for endian conversion of bfloat16 types.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=HDFGroup%2Fhdf5&utm_source=github&utm_medium=referral)<sup> for 3634271617213428fcc6fccda0e8c75d2b197c07. You can [customize](https://app.ellipsis.dev/HDFGroup/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

jhendersonHDF commented:
Essentially finished, pending CI results and some potential testing on a big-endian system

jhendersonHDF commented on doxygen/dox/DDLBNF200.dox:
```diff
@@ -93,10 +93,11 @@ This section contains a brief explanation of the symbols used in the DDL.
             H5T_IEEE_F32BE     | H5T_IEEE_F32LE     |
             H5T_IEEE_F64BE     | H5T_IEEE_F64LE     |
-            H5T_NATIVE_FLOAT16 | H5T_NATIVE_FLOAT   | 
+            H5T_FLT_BF16BE     | H5T_FLT_BF16LE     |
```
Open to any alternative names that seem more fitting, but this type is distinct from the IEEE standard, so I basically created a new category for alternative floating-point formats.

ajelenak replied:
Perhaps "FLOAT" instead of "FLT"?

jhendersonHDF replied:
Seems reasonable to me

qkoziol replied:
Agree - I like "FLOAT" better also

qkoziol replied:
Actually, maybe "NONSTD", since it's a contrast to the "IEEE" label?

ajelenak replied:
So... `H5T_NONSTD_BF16` or `H5T_NONSTD_BFLOAT16`? I like "FLOAT" explicitly.

jhendersonHDF replied:
The issue I see with that is that it's not unlikely that these types get adopted into some standard in the future, whether it's IEEE or not.

jhendersonHDF commented on src/H5Tnative.c:
```diff
@@ -739,17 +741,39 @@ H5T__get_native_float(size_t size, H5T_direction_t direction, size_t *struct_ali
+            if (NULL == (bf16_be_dt = H5I_object(H5T_FLT_BF16BE)))
+                HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, NULL, "not a data type");
+
+            /* Promote bfloat16 to float instead of float16, as it
```
bfloat16 types should be promoted to `float` instead of float16, as the type is the same size as float16, but a different format. Converting between bfloat16 and `float` is also very simple (by design).

fortnern replied:
Is the reasoning here that you don't want to lose data by shrinking the exponent? Wouldn't that apply to (almost) all 16 bit floats that aren't IEEE?

jhendersonHDF replied:
Yes, float should be the closest native type that can represent data correctly and bfloat16's design makes conversion to/from IEEE 32-bit floats simple. This should be revisited when support for a native bfloat16 type is added though, as there would be no need for promotion at that point. I suppose that, yes, there are some formats you could choose for a 16-bit floating point number where this logic promoting them to the _Float16 format instead would be problematic. Happy to fix this to just promote everything of this size that isn't _Float16 to float or even do some more format checking to decide the right type to promote to, but there are several other places in the library like that which I believe we should address as well.

jhendersonHDF commented on tools/lib/h5tools_type.c:
```diff
@@ -55,8 +55,12 @@ h5tools_get_little_endian_type(hid_t tid)
-            if (size == 2)
-                p_type = H5Tcopy(H5T_IEEE_F16LE);
+            if (size == 2) {
+                if (true == H5Tequal(tid, H5T_IEEE_F16LE) || true == H5Tequal(tid, H5T_IEEE_F16BE))
```
Had to make sure the correct type between float16 and bfloat16 is picked so that the data comes out correctly.

fortnern replied:
Similarly, seems this could cause problems with custom floats in general. No need to fix for this PR though.

jhendersonHDF replied:
Yes, from the other comment there are several places, especially in the tools and Java code, where we should handle such types better. Handling of those types has been poor, even outside of my changes here which admittedly do introduce problems with specific custom 16-bit floating point formats.

jhendersonHDF replied:
For now, leaving the cases in this file and `java/src/jni/h5util.c` unresolved as the functions would need to be re-written quite a bit to handle custom types properly. But since these functions only return a copy of the related type if the input type matches one of the specific types listed here, custom types should be no worse off than before. For an unmatched type, the functions will just return `H5I_INVALID_HID` which seems better than picking an incorrect type.

jhendersonHDF commented on tools/test/h5dump/h5dumpgentest.c:
```diff
@@ -12753,6 +12773,156 @@ gent_complex_be(void)
+    hsize_t dims[2], adims[1];
+
+    /*
+     * bfloat16 keeps approximately the same range as the IEEE 32-bit
```
Eventually better tests can be written, but for now this PR just adds support for predefined types and doesn't add support for a native type. GCC and Clang both have support for a __bf16 type at this point though, so it should be doable in the future.

qkoziol commented:
I'll try to review tomorrow and get you some feedback

epourmal commented:
It was my undesrtanding that middle part "_IEEE_" or similar represents
architecture or a standard. Should we follow the same rule here?

On Mon, Mar 24, 2025 at 12:11 PM Quincey Koziol ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In doxygen/dox/DDLBNF200.dox
> <https://github.com/HDFGroup/hdf5/pull/5402#discussion_r2010607578>:
>
> >              H5T_IEEE_F32BE     | H5T_IEEE_F32LE     |
>              H5T_IEEE_F64BE     | H5T_IEEE_F64LE     |
> -            H5T_NATIVE_FLOAT16 | H5T_NATIVE_FLOAT   |
> +            H5T_FLT_BF16BE     | H5T_FLT_BF16LE     |
>
> Agree - I like "FLOAT" better also
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/HDFGroup/hdf5/pull/5402#discussion_r2010607578>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADLFT3POWBKSASYVAQ4VSTD2WA4CRAVCNFSM6AAAAABZRCFMH2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDOMJRGEZTANZUHA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

qkoziol commented:
> It was my undesrtanding that middle part "_IEEE_" or similar represents architecture or a standard. Should we follow the same rule here?
> […](#)
> On Mon, Mar 24, 2025 at 12:11 PM Quincey Koziol ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In doxygen/dox/DDLBNF200.dox <[#5402 (comment)](https://github.com/HDFGroup/hdf5/pull/5402#discussion_r2010607578)>: > H5T_IEEE_F32BE | H5T_IEEE_F32LE | H5T_IEEE_F64BE | H5T_IEEE_F64LE | - H5T_NATIVE_FLOAT16 | H5T_NATIVE_FLOAT | + H5T_FLT_BF16BE | H5T_FLT_BF16LE | Agree - I like "FLOAT" better also — Reply to this email directly, view it on GitHub <[#5402 (comment)](https://github.com/HDFGroup/hdf5/pull/5402#discussion_r2010607578)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADLFT3POWBKSASYVAQ4VSTD2WA4CRAVCNFSM6AAAAABZRCFMH2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDOMJRGEZTANZUHA> . You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

Right - that was why I thought "NONSTD" might be better

qkoziol commented:
Maybe "NONSTD_BFLOAT16" ?

jhendersonHDF commented:
> It was my undesrtanding that middle part "_IEEE_" or similar represents architecture or a standard. Should we follow the same rule here?

I'm not sure that this is necessarily the case and I'd argue we should consider abandoning that. For example, `H5T_STD_B32BE` doesn't really mean anything, because AFAIK there's really no standard type for this. C23 does appear to introduce those, but we don't target C23 yet.

jhendersonHDF commented:
> Maybe "NONSTD_BFLOAT16" ?

That name is a bit weird to me because bfloat16 **is the standard**. There's no other standard for the type, it's just that the standard isn't an IEEE standard.

jhendersonHDF commented:
Also consider I plan to add support for FP8, FP6 and FP4 following this PR, and in each case the formats essentially are the standard until they have wider adoption. https://arxiv.org/abs/2209.05433 for example.

qkoziol commented:
Since Google created it, maybe "GOOGLE_BFLOAT16" ?

ajelenak commented:
I was thinking the same...

epourmal commented:
> Also consider I plan to add support for FP8, FP6 and FP4 following this PR, and in each case the formats essentially are the standard until they have wider adoption. https://arxiv.org/abs/2209.05433 for example.

Frankly speaking, I would drop middle part for these types and document their implementation. I would vote against introducing GOOGLE.

jhendersonHDF commented:
> Since Google created it, maybe "GOOGLE_BFLOAT16" ?

Perhaps. The reason I used BF16 on the end part was just to match our existing conventions like `H5T_IEEE_F32LE`, especially because there will be types like F8E4M3 (FP8 with 4 exponent bits and 3 mantissa bits). That said, I also like FLOAT being spelled out explicitly, but I still think it's much more natural to just use `H5T_FLOAT_` so that we don't tie the names to any particular standard, architecture or company unless it makes sense to. I see this as "one datatype of class `H5T_FLOAT` is called `H5T_FLOAT_BF16`". I'd definitely like to avoid the `STD` vs. `NONSTD` naming if possible, because I'm fairly certain these types will be adopted into some standard in the future.

ajelenak commented:
I was ok with `H5T_FLOAT_BF16` to begin with but enjoy discussing alternative naming options.

jhendersonHDF commented:
For the time being, I'm going to proceed on the other datatypes using `H5T_FLOAT_` names until we come to a resolution.

I propose that we use a convention of `H5T_<TYPE_CLASS>_<OPTIONAL(?)_ARCH_OR_STANDARD>_<SPECIFICS>` going forward and welcome any feedback if others think differently. This leads to names like `H5T_FLOAT_INTEL_F32`, `H5T_FLOAT_IEEE_F32LE`, `H5T_INTEGER_I32BE`, etc. The question mark on optional is because I'm thinking it might be worth it to always mention a standard or architecture to be consistent, leading to, for example, `H5T_INTEGER_C_I32BE` for `int32_t`. Note that this is already the convention I used for predefined complex datatypes (`H5T_COMPLEX_IEEE_F32LE`, for example). I believe this form does a better job of actually distinguishing what a particular datatype is. For example, `H5T_IEEE_F32LE` tells you that the datatype comes from or relates to an IEEE standard, but you rely on the "F32" part to tell you that it's a floating-point type. If additional non-floating-point types are added to IEEE standards, this may get confusing. For example, if we wanted to add support for the decimal floating-point types, I imagine we'd currently use something like `H5T_IEEE_D32LE` (maybe `H5T_IEEE_DF32LE`). However, that "D" would be somewhat in conflict with the current `H5T_UNIX_D32LE` and related types, where "D" apparently represents time data. If an IEEE standard included time types (for whatever reason), this would just complicate things more. A different letter than "D" could of course be picked to represent decimal floating-point datatypes, but I feel adding the type class gives context to anything in the "specifics" part of the name and helps deal with any conflict between lettering. It may also be worth revisiting what exactly should be in the "specifics" part of the name.

epourmal commented:
Makes sense, but I would leave old types as they are and use new standard only for the new types.

jhendersonHDF commented:
> Makes sense, but I would leave old types as they are and use new standard only for the new types.

I certainly wouldn't want to go changing old type names as it would be nothing but a source of annoyance (though we can always introduce new macros that just point to the old names if we wanted).

fortnern commented:
I don't think the C standard specifies the exact integer format, or even that it's two's complement, so we shouldn't name types "C". I'm not sure there's a better name for "normal" integer types than STD, but I might be wrong. Adding new names for consistency is fine, and something like H5T_INTEGER_U32LE is probably fine, though H5T_INTEGER_I32LE seems redundant. Maybe _S32LE or just _32LE?

jhendersonHDF commented:
> or even that it's two's complement

Note that this is the case in C23 at least.

> so we shouldn't name types "C"

I tend to agree, it's just that STD doesn't really particularly mean anything.

> though H5T_INTEGER_I32LE seems redundant.

Agree. Also why I mentioned we may want to revisit what should go in the "specifics" part of the name if we use the convention. FLOAT_XXX_F32LE is just as redundant, so maybe the single letter on the end of these isn't needed and we should consider something else?

brtnfld commented:
Jordan, is this a summary of what you are proposing:
```

H5T_<TYPE_CLASS>_<OPTIONAL_QUALIFIER>_<SPECIFICS>

    H5T_<TYPE_CLASS>: This explicitly states the data type (e.g., FLOAT, INTEGER, COMPLEX).
    _<OPTIONAL_QUALIFIER>: This qualifier provides additional context, such as:
        Architecture: (e.g., INTEL)
        Specific Organization: (e.g. GOOGLE)
        Standard: (e.g. IEEE)
        This section can be omitted if the datatype is very common or there is no specific qualifier.
    _<SPECIFICS>: This details the specific format (e.g., BF16, F32LE, I32BE, F8E4M3).
```

jhendersonHDF commented:
> Jordan, is this a summary of what you are proposing:
> 
> H5T_<TYPE_CLASS>_<OPTIONAL_QUALIFIER>_
> 
> ```
> H5T_<TYPE_CLASS>: This explicitly states the data type (e.g., FLOAT, INTEGER, COMPLEX).
> _<OPTIONAL_QUALIFIER>: This qualifier provides additional context, such as:
>     Architecture: (e.g., INTEL)
>     Specific Organization: (e.g. GOOGLE)
>     Standard: (e.g. IEEE)
>     This section can be omitted if the datatype is very common or there is no specific qualifier.
> _<SPECIFICS>: This details the specific format (e.g., BF16, F32LE, I32BE, F8E4M3).
> ```

Yes, this looks like a good summary. Thanks!

brtnfld commented:
@ellipsis-dev, review PR.

mattjala commented on tools/src/h5ls/h5ls.c:
```diff
@@ -598,6 +599,12 @@ print_ieee_type(h5tools_str_t *buffer, hid_t type, int ind)
+    else if (H5Tequal(type, H5T_FLOAT_BFLOAT16BE_g) == true) {
+        h5tools_str_append(buffer, "bfloat16 16-bit big-endian float");
+    }
+    else if (H5Tequal(type, H5T_FLOAT_BFLOAT16LE_g) == true) {
```
Is there a reason to use the `_g` values directly here instead of the macros from `H5Tpublic.h` like for the other types?

jhendersonHDF replied:
Was probably a find/replace issue. No problem using them directly, but I'll fix this

fortnern commented on fortran/src/H5_f.c:
```diff
@@ -245,6 +245,14 @@ h5init_types_c(hid_t_f *types, hid_t_f *floatingtypes, hid_t_f *integertypes)
         return ret_value;
     if ((floatingtypes[3] = (hid_t_f)H5Tcopy(H5T_IEEE_F64LE)) < 0)
         return ret_value;
+    if ((floatingtypes[4] = (hid_t_f)H5Tcopy(H5T_IEEE_F16BE)) < 0)
```
Were the IEEE 16 bit types missing here before?

jhendersonHDF replied:
Yes, I could add those in a separate PR if need be. They simply weren't added at the time support for _Float16 was introduced, but as this PR adds in a 16-bit floating point type it seemed reasonable to add them now.

fortnern commented:
Looks good apart from a few minor comments (2 in replies) (and Matt's)

jhendersonHDF commented:
Marked as draft again until remaining comments can be addressed.

jhendersonHDF committed:
Add predefined datatypes for bfloat16 data

Adds predefined datatypes for little- and big-endian bfloat16 data

Does not add support for any native bfloat16 types; datatype conversions are performed in software

Also adds missing float16 predefined types to fortran

lrknox committed:
Merge branch 'develop' into feature/bfloat16_predefined

lrknox commented:
Approvals from fortnern and brtnfld were dismissed by resolution of CHANGELOG.md conflict.
